mitiq/mitiq_pyquil/executor.py

```diff
@@ -0,0 +1,105 @@
+# Copyright (C) 2020 Unitary Fund
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+"""Executor templates for working with pyQuil.
+"""
+
+from typing import Callable
+
+import numpy as np
+from pyquil import Program
+from pyquil.api import QuantumComputer
+from pyquil.experiment import bitstrings_to_expectations
+from pyquil.gates import MEASURE, RESET
+
+from mitiq.mitiq_pyquil.compiler import basic_compile
+
+
+def generate_qcs_executor(
+    qc: QuantumComputer,
+    expectation_fn: Callable[[np.ndarray], float],
+    shots: int = 1000,
+    reset: bool = True,
+    debug: bool = False,
+) -> Callable[[Program], float]:
+    """
+    Generates an executor for QCS that ingests pyQuil programs.
+
+    :param qc: The QuantumComputer object to use as backend.
+    :param expectation_fn: Takes in bitstring results and produces a float.
+    :param shots: Number of shots to take.
+    :param reset: Whether or not to enable active reset.
+    :param debug: If true, print the program after compilation.
+    """
+
+    def executor(program: Program) -> float:
+        p = Program()
+
+        # add reset
+        if reset:
+            p += RESET()
+
+        # add main body program
+        p += program.copy()
+
+        # add memory declaration
+        qubits = p.get_qubits()
+        ro = p.declare("ro", "BIT", len(qubits))
+
+        # add measurements
+        for idx, q in enumerate(qubits):
+            p += MEASURE(q, ro[idx])
+
+        # add numshots
+        p.wrap_in_numshots_loop(shots)
+
+        # nativize the circuit
+        p = basic_compile(p)
+
+        # print out nativized program
+        if debug:
+            print(p)
+
+        # compile the circuit
+        b = qc.compiler.native_quil_to_executable(p)
+
+        # run the circuit, collect bitstrings
+        qc.reset()
+        results = qc.run(b)
+
+        # compute expectation value
+        return expectation_fn(results)
+
+    return executor
+
+
+def ground_state_expectation(results: np.ndarray) -> float:
+    """
+    Example expectation_fn. Computes the ground state expectation, also
+    called survival probability.
+    """
+    num_shots = len(results)
+    return (
+        num_shots - np.count_nonzero(np.count_nonzero(results, axis=1))
+    ) / num_shots
+
+
+def zz_expectation(results: np.ndarray) -> float:
```

Comment:
Is this just for two qubit circuits (measurement data)? Should it be generalized to Zi Zj for any pair (i, j) of qubits?

Reply:
This file is really for templates & examples -- I don't think it needs to be generalized necessarily.

Reply:
Since the ground state function provides both a template and a (general) example, I would remove `zz_expectation`. I'll leave it to your discretion, though.